File: src/main/java/tafqeetj/numbers/NumberSign.java
package tafqeetj.numbers;

import tafqeetj.converters.BasicNumberConverter;
import tafqeetj.converters.BasicNumberConverterFactory;
import tafqeetj.converters.QuantitiveNumberConverterFactory;
import tafqeetj.converters.negative.NegativeQuantitiveNumberConvertersFactory;
import tafqeetj.converters.positive.PositiveQuantitiveNumberConvertersFactory;

import java.math.BigDecimal;
import java.math.BigInteger;

public enum NumberSign {
    POSITIVE(
            PositiveQuantitiveNumberConvertersFactory.getInstance(),
            BasicNumberConverterFactory.getPositiveConverter()
    ),
    NEGATIVE(
            NegativeQuantitiveNumberConvertersFactory.getInstance(),
            BasicNumberConverterFactory.getNegativeConverter()
    );

    private final QuantitiveNumberConverterFactory quantitiveNumberConverterFactory;
    private final BasicNumberConverter basicNumberConverter;

    NumberSign(QuantitiveNumberConverterFactory quantitiveNumberConverterFactory,
               BasicNumberConverter basicNumberConverter) {
        this.quantitiveNumberConverterFactory = quantitiveNumberConverterFactory;
        this.basicNumberConverter = basicNumberConverter;
    }

    public QuantitiveNumberConverterFactory getQuantitiveNumberConverterFactory() {
        return quantitiveNumberConverterFactory;
    }

    public BasicNumberConverter getBasicNumberConverter() {
        return basicNumberConverter;
    }

    public static NumberSign signOf(int number) {
        if (number >= 0)
            return POSITIVE;
        return NEGATIVE;
    }

    public static NumberSign signOf(long number) {
        if (number >= 0L)
            return POSITIVE;
        return NEGATIVE;
    }

    public static NumberSign signOf(short number) {
        if (number >= 0)
            return POSITIVE;
        return NEGATIVE;
    }

    public static NumberSign signOf(byte number) {
        if (number >= 0)
            return POSITIVE;
        return NEGATIVE;
    }

    public static NumberSign signOf(BigInteger number) {
        if (number.signum() >= 0)
            return POSITIVE;
        return NEGATIVE;
    }

    public static NumberSign signOf(BigDecimal number) {
        if (number.signum() >= 0)
            return POSITIVE;
        return NEGATIVE;
    }
}

File: src/main/kotlin/tech/moussa/tafqeetk/numbers/NumberSign.kt
package tech.moussa.tafqeetk.numbers

import tech.moussa.tafqeetk.converters.BasicNumberConverter
import tech.moussa.tafqeetk.converters.BasicNumberConverterFactory
import tech.moussa.tafqeetk.converters.QuantitativeNumberConverterFactory
import tech.moussa.tafqeetk.converters.negative.NegativeQuantitativeNumberConvertersFactory
import tech.moussa.tafqeetk.converters.positive.PositiveQuantitativeNumberConvertersFactory

import java.math.BigDecimal
import java.math.BigInteger

enum class NumberSign(
        val value: QuantitativeNumberConverterFactory,
        val numberConverter: BasicNumberConverter
) {
    POSITIVE(
            PositiveQuantitativeNumberConvertersFactory.getInstance(),
            BasicNumberConverterFactory.positiveConverter
    ),
    NEGATIVE(
            NegativeQuantitativeNumberConvertersFactory.getInstance(),
            BasicNumberConverterFactory.negativeConverter
    );

    companion object {
        fun signOf(number: Int): NumberSign {
            return if (number >= 0) POSITIVE else NEGATIVE
        }

        fun signOf(number: Long): NumberSign {
            return if (number >= 0L) POSITIVE else NEGATIVE
        }

        fun signOf(number: Short): NumberSign {
            return if (number >= 0) POSITIVE else NEGATIVE
        }

        fun signOf(number: Byte): NumberSign {
            return if (number >= 0) POSITIVE else NEGATIVE
        }

        fun signOf(number: BigInteger): NumberSign {
            return if (number.signum() >= 0) POSITIVE else NEGATIVE
        }

        fun signOf(number: BigDecimal): NumberSign {
            return if (number.signum() >= 0) POSITIVE else NEGATIVE
        }
    }
}